src/env.rs

```diff
@@ -0,0 +1,227 @@
+        }
+        self.virt_opts.env.as_mut().unwrap()
+    }
+
+    pub fn env_host_allow(mut self, allow_list: &[&str]) -> Self {
+        let env = self.get_or_create_env();
+        env.host = HostEnv::Allow(allow_list.iter().map(|s| s.to_string()).collect());
+        self
+    }
+
+    pub fn env_host_deny(mut self, deny_list: &[&str]) -> Self {
+        let env = self.get_or_create_env();
+        env.host = HostEnv::Deny(deny_list.iter().map(|s| s.to_string()).collect());
+        self
+    }
+
+    pub fn env_host_all(mut self) -> Self {
+        let env = self.get_or_create_env();
+        env.host = HostEnv::All;
+        self
+    }
+
+    pub fn env_host_none(mut self) -> Self {
+        let env = self.get_or_create_env();
+        env.host = HostEnv::None;
+        self
+    }
+
+    pub fn env_overrides(mut self, overrides: &[(&str, &str)]) -> Self {
+        let env = self.get_or_create_env();
+        for (key, val) in overrides {
+            env.overrides.push((key.to_string(), val.to_string()));
+        }
+        self
+    }
+}
+
+pub fn create_env_virt<'a>(module: &'a mut Module, env: &VirtEnv) -> Result<()> {
+    let env_ptr_addr = {
+        let env_ptr_export = module
+            .exports
+            .iter()
+            .find(|expt| expt.name.as_str() == "get_env_ptr")
+            .unwrap();
+        let ExportItem::Function(func) = env_ptr_export.item else {
+            panic!()
+        };
+        let FunctionKind::Local(local_func) = &module.funcs.get(func).kind else {
+            panic!()
+        };
+        let func_body = local_func.block(local_func.entry_block());
+        if func_body.instrs.len() != 1 {
+            return Err(anyhow!(
+                "Unexpected get_env_ptr implementation. Should be a constant address return."
+            ));
+        }
+        let Instr::Const(Const {
+            value: Value::I32(env_ptr_addr),
+        }) = &func_body.instrs[0].0
+        else {
+            return Err(anyhow!(
+                "Unexpected get_env_ptr implementation. Should be a constant address return."
+            ));
+        };
+        *env_ptr_addr as u32
+    };
+
+    // If host env is disabled, remove its import entirely
+    // replacing it with a stub panic
+    if matches!(env.host, HostEnv::None) {
+        stub_imported_func(module, "wasi:cli-base/environment", "get-environment");
+        // we do arguments as well because virt assumes reactors for now...
+        stub_imported_func(module, "wasi:cli-base/environment", "get-arguments");
+    }
+
+    let memory = module.memories.iter().nth(0).unwrap().id();
+
+    // prepare the field data list vector for writing
+    // strings must be sorted as binary searches are used against this data
+    let mut field_data_vec: Vec<&str> = Vec::new();
+    for (key, value) in &env.overrides {
+        field_data_vec.push(key.as_ref());
+        field_data_vec.push(value.as_ref());
+    }
+    field_data_vec.sort();
```

Comment:
Is this `sort` call correct? Given that we're not storing this as `(&str, &str)`, the sorting would shuffle the keys and the values of the overrides, right? From what I can tell, only the allow/deny list needs to be sorted.

Reply:
This was a bug, and a huge catch. Thank you!